File: solutions/084.Largest_Rectangle_in_Histogram/AC_stack_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_stack_n.java
*  Create Date: 2015-03-07 13:34:35
*  Descripton:  Using increase-only-stack to deal with it.
*/

import java.util.*;

public class Solution {

    public int largestRectangleArea(int[] height) {
        int len = height.length;
        Stack<Integer> stk = new Stack<Integer>();
        int ret = 0;
        for (int i = 0; i <= len; ++i) {
            int h = (i == len ? 0 : height[i]);
            if (stk.isEmpty() || h >= height[stk.peek()]) {
                stk.push(i);
            } else {
                int top = stk.pop();
                ret = Math.max(ret, height[top] * (stk.empty() ? i : i - stk.peek() - 1));
                --i;    // back one step again
            }
        }
        return ret;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 4};
        System.out.println("no case");
    }
}

File: solutions/020.Valid_Parentheses/AC_stack_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_stack_n.cpp
*  Create Date: 2014-12-05 14:19:58
*  Descripton:  use stack
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        int len = s.length();
        for (int i = 0; i < len; i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                stk.push(s[i]);
            } else {
                if (stk.empty())
                    return false;
                if (stk.top() == '(' && s[i] == ')')
                    stk.pop();
                else if (stk.top() == '[' && s[i] == ']')
                    stk.pop();
                else if (stk.top() == '{' && s[i] == '}') 
                    stk.pop();
                else
                    return false;
            }
        }
        return stk.empty();
    }
};

int main() {
    Solution a;
    string s;
    while (cin >> s)
        cout << a.isValid(s) << endl;
    return 0;
}

